Improve evaluation of nested `ComposedFunction`s

martinholters opened the pull request:
* ~~Add (the equivalent of) `@assume_effects :terminates_globally` to `unwrap_composed`. Although it could be inferred as `Const`, without the annotation, it was not elided for too complex inputs, resulting in unnecessary runtime overhead.~~
  EDIT: now #45993 is merged and this part isn't necessary.
* Reverse recursion order in `call_composed`. This prevents potentially changing argument types being piped through the recursion, making inference bail out. With the reversed order, only the tuple of remaining functions is changing during recursion and is becoming strictly simpler, letting inference succeed.

```julia
julia> f = reduce(∘, fill(sin,500));

julia> const g = reduce(∘, fill(sin,500));

julia> @time f(1.0)
  1.474751 seconds (1.83 M allocations: 120.273 MiB, 3.22% gc time, 99.00% compilation time) # master
  6.011575 seconds (4.02 M allocations: 278.036 MiB, 3.33% gc time, 100.00% compilation time) # PR
0.07698641344541407

julia> @btime f(1.0)
  14.102 ms (1000 allocations: 15.62 KiB) # master
  5.337 μs (1 allocation: 16 bytes) # PR
0.07698641344541407

julia> @btime g(1.0)
  14.091 ms (998 allocations: 15.59 KiB) # master
  1.878 ns (0 allocations: 0 bytes) # PR
0.07698641344541407
```
So increase in compilation time is noticeable, but execution is also much faster. (Yes, we can const-prop and completely elide `g` here). Also note that compilation time is still much better than with 1.8-rc1 (or 1.7), where it is several minutes.

That said, the fact that we have to replace the concise `(c::ComposedFunction)(x...; kw...) = c.outer(c.inner(x...; kw...))` with the convoluted construction still makes me sad.

Fixes #45877.

martinholters committed:
Improve evaluation of nested `ComposedFunction`s

* Add (the equivalent of) `@assume_effects :terminates_globally` to
  `unwrap_composed`. Although it could be inferred as `Const`, without
  the annotation, it was not elided for too complex inputs, resulting in
  unnecessary runtime overhead.
* Reverse recursion order in `call_composed`. This prevents potentially
  changing argument types being piped through the recursion, making
  inference bail out. With the reversed order, only the tuple of
  remaining functions is changing during recursion and is becoming
  strictly simpler, letting inference succeed.

N5N3 commented on base/operators.jl:
```diff
@@ -1019,12 +1019,20 @@ end
 function (c::ComposedFunction)(x...; kw...)
     fs = unwrap_composed(c)
-    call_composed(fs[1](x...; kw...), tail(fs)...)
+    fs[1](call_composed(tail(fs), x, kw))
```
There seems no need to split `fs[1]` with reversed recursion.
```julia
(c::ComposedFunction)(x...; kw...) = call_composed(unwrap_composed(c), x, kw)
```

aviatesk commented on base/operators.jl:
```diff
@@ -1019,12 +1019,20 @@ end
+    fs[1](call_composed(tail(fs), x, kw))
 end
-unwrap_composed(c::ComposedFunction) = (unwrap_composed(c.inner)..., unwrap_composed(c.outer)...)
+eval(quote
+    # @assume_effects :terminates_globally, but @assume_effects cannot be used
+    # Core.Compiler, hence the Expr(:meta, ...)
+    function unwrap_composed(c::ComposedFunction)
+        #$(Expr(:meta, :($(Expr(:purity, false, false, false, true, false, false)))))
+        $(Expr(:meta, Expr(:purity, false, false, false, true, false, false)))
+        return (unwrap_composed(c.outer)..., unwrap_composed(c.inner)...)
+    end
+end)
```
I think we generally want to avoid direct creation of `:purity` expression in this way as it is fragile against future changes in the effect system design.

I'd like to add this diff:
```diff
diff --git a/base/essentials.jl b/base/essentials.jl
index 906c36ad9c..8eee2923b3 100644
--- a/base/essentials.jl
+++ b/base/essentials.jl
@@ -204,23 +204,35 @@ macro _pure_meta()
 end
 # can be used in place of `@assume_effects :total` (supposed to be used for bootstrapping)
 macro _total_meta()
-    return _is_internal(__module__) && Expr(:meta, Expr(:purity,
-        #=:consistent=#true,
-        #=:effect_free=#true,
-        #=:nothrow=#true,
-        #=:terminates_globally=#true,
-        #=:terminates_locally=#false,
-        #=:notaskstate=#true))
+    return _is_internal(__module__) && Expr(:meta, _assumed_effects(;
+        consistent=true,
+        effect_free=true,
+        nothrow=true,
+        terminates_globally=true,
+        notaskstate=true))
 end
 # can be used in place of `@assume_effects :foldable` (supposed to be used for bootstrapping)
 macro _foldable_meta()
-    return _is_internal(__module__) && Expr(:meta, Expr(:purity,
-        #=:consistent=#true,
-        #=:effect_free=#true,
-        #=:nothrow=#false,
-        #=:terminates_globally=#true,
-        #=:terminates_locally=#false,
-        #=:notaskstate=#false))
+    return _is_internal(__module__) && Expr(:meta, _assumed_effects(;
+        consistent=true,
+        effect_free=true,
+        terminates_globally=true))
+end
+
+function _assumed_effects(;
+    consistent::Bool = false,
+    effects_free::Bool = false,
+    nothrow::Bool = false,
+    terminates_globally::Bool = false,
+    terminates_locally::Bool = false,
+    notaskstate::Bool = false)
+    return Expr(:purity,
+        consistent,
+        effects_free,
+        nothrow,
+        terminates_globally,
+        terminates_locally,
+        notaskstate)
 end
 
 # another version of inlining that propagates an inbounds context
```

and then rewrite this piece of code like:
```suggestion
eval(quote
    # equivalent to `@assume_effects :terminates_globally`,
    # but @assume_effects cannot be used `Core.Compiler`, hence `_assumed_effects(...)`
    function unwrap_composed(c::ComposedFunction)
        $(Expr(:meta, _assumed_effects(; terminates_globally=true)))
        return (unwrap_composed(c.outer)..., unwrap_composed(c.inner)...)
    end
end)
```

Then we don't look for all the usages of `:purity` expression when we add new effect to the system as far as we keep the same naming schema for the existing effects.

aviatesk replied:
It seems more difficult than I expected to use kwfunc until we define namedtuple.jl.. We can leave this to the future and don't need to let it hold this PR.

martinholters replied:
Alternatively, we could move the `ComposedFunction` stuff to a dedicated file that is included after `array.jl` (which makes `@assume_effects` work) but before `bitarray.jl` (which uses composition with `!iszero`).

aviatesk commented:
You can test this improvement as like:
> test/operators.jl
```julia
include("compiler/irutils.jl")
const issue_45877 = reduce(∘, fill(sin,500))
@test fully_eliminated() do
    issue_45877(1.0)
end
```

martinholters committed:
Move `ComposedFunction` to dedicated file

Motivation is to be able to use `@assume_effects` by including that file
at an appropriate position during bootstrap.

martinholters commented:
@aviatesk can you explain why the `@assume_effects` is necessary? On master I find
```julia
julia> const g = reduce(∘, fill(sin,500));

julia> only(code_typed() do
           Base.unwrap_composed(g)
       end)[2]
NTuple{500, typeof(sin)}

julia> Base.infer_effects() do
           Base.unwrap_composed(g)
       end
(+c,+e,+n,!t,+s)
```
So inference succeeds which implies that it follows the recursion to the end, therefore obviously has figured out that there is an end. But the effects don't seem to catch that. Is there some way we could make the effects here more precise? If we could avoid the `@assume_effects`, there would also no longer be a need for the code movement.

aviatesk reviewed (approved):
We can try to improve the recursion detection for effect analysis instead, but I'm also fine with merging this for now and tackle the improvement later.

martinholters commented:
I don't mind either way. I won't have the bandwidth to look at improving effect analysis anytime soon, don't know whether you can tackle it in the foreseeable future. If no, we probably should merge as is. But I have no personal interest in deeply nested composed functions, so no hurry from my side.

ianatol commented:
LGTM.

I think the `@assume_effects` is fine for the time being, while Shuhei looks into improving the recursion detection for effects.

aviatesk committed:
Revert "Move `ComposedFunction` to dedicated file"

This reverts commit bb5ea07319a47d325b57b7b62c561bd6fa325838.

aviatesk committed:
remove the unncessary effect annotation from `unwrap_composed`
